### ISSUE TITLE
Enhancement: [unbound-method] allow assigment as a property of an object of the same type

### ISSUE THREAD

**Issue by miguel-leon** [typescript-eslint/typescript-eslint#10427]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/unbound-method/

### Description

If the unbound method is being assigned as to a property of an object of the same type, the method doesn't need to be bound, because it will then also have a `this` of the same type.

### Fail

```TypeScript
class MyClass {
  foo() {
    this.foo = this.bar;
  }

  bar() {

  }
}
```

### Pass

```TypeScript
N/A
```

### Additional Info

[playground](https://typescript-eslint.io/play/#ts=5.7.2&fileType=.ts&code=MYGwhgzhAECyCeBhcVoG8BQ1oDMD2eAFAJTpbbQAuAFgJYQB0%2Be0AvFXYwEZgBOA3OQC%2BGcj14kywjCKA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6WJgIwHsOATWgFtE%2BABbc%2B6MAG1IiaNG7RIAGkzhsOSgHMmSxAGV8AQ3yUyU-HEQaIAXwC6GuyDtA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

**Comment by kirkwaiblinger** [comment]
Hi @miguel-leon!

Could you provide an example of how this might be useful in organically occurring code? It's not obvious to me how this pattern might be used (i.e. the non-niche criterion). Especially where the `foo()` prototype method results in a `this.foo = ` instance assignment... is that a deliberate part of the example?

**Comment by miguel-leon** [comment]
Hi @kirkwaiblinger,
The way it happened to me, is that I needed a `load` method that behaved differently only the first time it was called. Given javascript capabilities and how prototype chain works, I decided to reassign the public `load` to a private one and then in that method delete the own property from the instance.
Something like the following:
```ts
class Foo {
    constructor() {
        this.load = this.firstLoad;
    }

    private firstLoad() {
        delete (this as MutatingFoo).load;
        // ..
    }

    public load() {
        //...
    }
}

interface MutatingFoo {
	load: Foo['load'] | undefined;
}
```

I think there might be other use cases, another example could be assigning methods on prototypes or objects with partially equal structures to reuse functionality, like imitating multiple inheritance.

In any case it just sounds sensible to me that if the object a method is being assigned to, has the same type as `this` in the original method, then there's simply no need to bind it.

I hope the reasoning is agreeable enough for you. ✌️
Thanks!

**Comment by bradzacher** [comment]
We would still error in that example because the type of `this` and `this as MutatingFoo` are different types.
It's possible we could handle it if the assignment target is a member expression and the object of the expression is the same as the value's then we could exclude it.

But I too don't really see the real-world applicability. I don't really think that reassigning method implementations is a common usecase. Most codebases treat methods as purely readonly.

**Comment by kirkwaiblinger** [comment]
An example we could more readily analyze would look more like this
[playground](https://typescript-eslint.io/play/#ts=5.7.2&fileType=.ts&code=MYGwhgzhAECyCeBhcVoG8BQ1rDCEARmMANYBc0AFAC4AWAlhBQspBAJTQC8AfNAG4B7egBNu0OowB0EQQFsApgBEFAMzABXENQBCC2mH71BAJwDcGLNAAOJ%2BvzDUFNuw6cA1PBuddoAcllFaBE1TW0BLwUAQj8LKwgFakQ8QmISSmACChoGJjgkFA5uPiFRTkxsbEkIKVx8IlJxTItsAF9LbEDlUK1dfUNjE0pyq2xgQQA7WRAFKRBBAHMc6Vt7RwVPEG92Fuh2qxMFBOph9FGJXNqUhpJxapl5bvVevQMjUyt29qA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6WJgIwHsOATWgFtE%2BABbc%2B6MAG1IiaNG7RIAGkzhsOSgHMmSxAGV8AQ3yUyU-HEQaIAXwC6GuyDtA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDAruuGAL4g9A&tokens=false)
```ts
class MyClass {
  callback: (this: MyClass) => void = this.someDefaultBehavior;

  private privateValue = 'some default value!';

  setCallback(cb: (this: MyClass) => void) {
    this.callback = cb;
  }

  someDefaultBehavior() {
    console.log(this.privateValue);
  }

  reset() {
    this.callback = this.someDefaultBehavior
  }
}
```
... but then, I'd say - I'd read this much easier as 
```ts
class MyClass {
  private callback?: () => void;

  private privateValue = 'some default value!';

  execute() { 
    if (this.callback != null) {
      this.callback();
    } else {
      this.someDefaultBehavior();
    }
  }

  setCallback(cb: () => void) {
    this.callback = cb;
  }

  someDefaultBehavior() {
    console.log(this.privateValue);
  }

  reset() {
    this.callback = undefined;
  }
}
```

Your example is.... much more esoteric, in a way that's hard to analyze even if we did want to support this in concept, so I'm skeptical that it's a pattern that's valuable enough to warrant code to detect it robustly. But it is certainly is clever and fun! I would absolutely never have considered that approach 😁.

FWIW - I'd definitely rewrite your code as something like
```ts
class Foo {
  private isFirstLoad = true;

  public load() {
    if (this.isFirstLoad) {
      throw new Error('already loaded!')
    }
    this.isFirstLoad = false;

    //
  }

  // (optional) _if_ you want to expose whether it's safe to call load
  public canLoad(): boolean {
    return !this.isFirstLoad;
  }
}
```

**Comment by miguel-leon** [comment]
Hi guys! thanks for your responses!

> type of this and this as MutatingFoo are different types

Not really according to typescript, or at least type assertible. But the `delete` operation is another matter, @kirkwaiblinger posed a refactor where no `delete` is used, and in any case, the type assertion is only there to satisfy typescript because it shouldn't be necessary. Because of how methods work, deleting from `this` would never actually delete the method so it would never be undefined, it's just that typescript doesn't currently understands that.

> I too don't really see the real-world applicability

I am too in the real world you know 😅

> I'd say - I'd read this much easier as ... 
> ```ts
>  execute() { 
>    if (this.callback != null) {
>      this.callback();
>    } else {
>      this.someDefaultBehavior();
>    }
>  }
> ```

Sure, we can start a philosophical debate on software. I am aware, I assure you, that the most popular preference among programmers is to write things very "straightforward" or "beginner-friendly", specially in open source software or companies with large teams. And `if` statements that express intent so explicitly or literally, can certainly achieve that, even better would be `if (this.firstCallFlag) ... else ...`.

However, it is my believe that this preference is the result of a widespread intellectual laziness: If we're so (badly) accustomed of always relying on the same approach, it's no surprise we resist considering alternatives, and when presented, they inevitably seem alien or outrageous. Because of this, most programmers never develop the skills to quickly understand more complex code, relying on technologies in a thoughtless, robotic manner.

Returning to the example, I firmly believe that having less code to maintain—fewer entities like variables, methods and branches—pays off in the long run. And that's why I opt not to use flags, executors, setters, conditionals, run-time errors or any other approach that introduces unnecessary entities. I rather rely on the language's paradigm or capabilities, and javascript allows you to achieve the desired behavioral pattern with just the two methods `load` and `firstLoad`.

> is a recipe for disaster to assign methods to non-methods
> ```ts
>  public notAMethod = this.method
>  ...
>  const unsafe = foo.notAMethod;
> ```

Not really my case, and I don't understand why you would assign a method to a non-method without binding, but if you end up doing this, you can make it safe by typing it like below, and typescript would complain at the unsafe invokation.
```ts
public notAMethod: (this: Foo) => void = this.method
```

To conclude, just let me reiterate that I'm not doing any shenanigans here, the assignment is method to method, no intermediary untyped variables, the container object is the same type (it's even the same object), despite the `delete` the method is never even `undefined`, etc. This all look very sensible to me, so I'm guessing the analysis on all those degenerate edge cases is just for stress testing the suggestion, which is ok, but just only those examples and not mine, are the "bad code" ones.

I hope this clarifies any and all doubts. ✌️
Thanks again!

**Comment by bradzacher** [comment]
> Not really according to typescript, or at least type assertible

I don't think you understand.
You're thinking like a human.
Think how a machine might interpret this.

To TS it sees `Foo` on the left and `MutatingFoo` on the right.
These are two entirely distinct types -- one is a class and one is an interface.
Yes - one is assignable to one another -- but they're still distinct types.

How is a rule expected to reason about the code -- if it sees two distinct, unrelated types -- what should it do?

**Comment by miguel-leon** [comment]
@bradzacher why you stuck with that? The issue isn't even about that type assertion. You think I don't understand, I feel bad you didn't even listened to all I said and are just nit and picking on something else. And FYI, typescript treats types structurally, saying "one is a class and one is an interface" doesn't even apply.

To put it succinctly and bluntly, this issue is about the rule "unbound-method" reporting on a method that is not **_unbound_**, it still has `this`, it's the same type, it's the same object. No need to expect calling `.bind` and not need for a arrow function. That's it.

Is a rule option appropriate for that? That's the question.

Let's not argue here about how a rule would reason about something else like two "different" types.

**Comment by bradzacher** [comment]
> it still has this, it's the same type

No -- it does not.
Your right-hand-side function is declared as `load: Foo['load']`
Which means the type is `load(): void`.
TS does not store the `this` type on your signature nor does it store the `Foo['load']` relationship in the types -- it just stores the signature. You can see this in the [playground](https://typescript-eslint.io/play/#ts=5.6.2&showAST=types&fileType=.tsx&code=MYGwhgzhAEBiD29oG8CwAoa1jwHYQBcAnAV2APiIAoBKFDLLAgCwEsIA6EeMAE2gC80Fuw4AzVkUIAZHrwDcDaAF8MSgA5FWANzAEAptAlSCsvrXqZGvfSH0HoVETEjQAsiQJ7WuAOYJ4Gi45RSssAHpw6A4OJVV0DRIAIxBWYGhuczo0MOhImNirePiMHwMiMTBgQw8vAh9-REssTN4ALjhEAG0AclaegF1oAB9oElwbCVx9BQx4oA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYBfEWg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false):

![screenshot of types in playground](https://github.com/user-attachments/assets/4e89be9f-f84d-4a6c-9981-b2f7d93876ef)

In fact if you interrogate the types you will see that `Foo.load`'s type has ID 89, whereas `MutatingFoo.load`'s type has ID 91, and if you step into the union it's function signature has ID 90. This means that if the rule interrogated the types it would just see two completely distinct types.

There are all sorts of edge cases to consider around this. What does it mean to "have the same `this` context" or "have the same type"? What are the parameters we're looking for here?

> "one is a class and one is an interface" doesn't even apply

A class represents a type, another class might be a subclass of that class -- meaning that your example could be considered valid if `MutatingFoo` was a subclass of `Foo` because this would imply that the `this` context is "safe" to assign to.
We can interrogate and understand the heirarchy of classes easily via the type system which means the rule could understand this.

However in your example `MutatingFoo` is an interface -- which means that there's no possible hierarchical relationship. In fact an interface has no runtime representation at all and thus has no `this` context. So it's entirely relevant to mention this -- how can the rule go "the this context is the same" if one of the constructs doesn't even have a `this` context we can reliably reason about?

> Let's not argue here about how a rule would reason about something else like two "different" types.

Why not? This is entirely relevant to how we should assess this feature request.
As I've illustrated above -- the example you've provided proves this is something we need to understand in order to assess the viability of your request.

Based on the examples you've provided it looks like you're asking for something that's not possible because you're describing relationships we cannot programmatically reason about.

We need to understand how a rule might reason about this. This is part of the exchange of ideas so that we can understand if it's even possible to build what you're asking for.

**Comment by bradzacher** [comment]
> However, it is my believe that this preference is the result of a widespread intellectual laziness: If we're so (badly) accustomed of always relying on the same approach, it's no surprise we resist considering alternatives, and when presented, they inevitably seem alien or outrageous.

It's important to consider how common a pattern is.
We're a small team with very limited bandwidth.

Supporting every code pattern is not something we can feasibly do and we knowingly and purposefully do not support everything because we do not have the time to build support or maintain the logic to support everything.

As encoded in our maintenance guide:
https://typescript-eslint.io/maintenance/issues#-feature-requests
> For any feature request, make sure the requested support is either:
> ...
> - Relevant for most projects that would be using typescript-eslint
>
> We avoid features that:
> - Are only relevant for a minority of users, as they aren't likely worth the maintenance burden

Understanding your current pattern in the context of other, more prevalent patterns is important for us to gauge if this pattern is something *_most_* users run into or not or if it's a rare/niche pattern.

If it's the former, then this is a request we can entertain.
If it's the latter then we'd likely close this as outside of our scope.

**Comment by miguel-leon** [comment]
Do you realize that the rule error is in line 3 right? and that the interface and type assertion can be removed as part of the example, as stated before, that it's just for illustration of an "organic" "real-world" scenario.

Put the original example in the playground
```ts
class MyClass {
  foo() {
    this.foo = this.bar;
  }

  bar() {}
}
```
and see if there's any type difference.

I don't think there's any need to look beyond an option that could be simply somehing like `allowMethodAssigment`.

If you don't agree, cool.
I don't see how arguing about all other not-so-related stuff is helping.

**Comment by bradzacher** [comment]
Consider the expansion of your examples:

```ts
class Foo {
  method() {}

  reassign1(other: Foo) {
    this.method = other.method;
    // is this valid?
  }
  reassign2(other: Foo) {
    other.method = this.method;
    // is this valid?
  }

  reassign3(other: SubFoo) {
    this.method = other.method;
    // is this valid?
  }
  reassign4(other: SubFoo) {
    other.method = this.method;
    // is this valid?
  }

  reassign5(other: IFoo) {
    this.method = other.method;
    // is this valid?
  }
  reassign6(other: IFoo) {
    other.method = this.method;
    // is this valid?
  }
}

class SubFoo extends Foo {}
interface IFoo {
  method: Foo['method'];
}
```

This is why these discussions are important. Where is the line with what we would support? How big is this request? Is it limited to cases that simply have `this.` on both sides? What about cases where the types are the same? Etc as shown above.

Currently your request is unbounded. I'm trying to tease out the bounds so that we can scope the size and understand the resulting cost vs value as a team.

This is why I say "think like a machine" - think how TS might evaluate these things and consider the cases that could arise. Consider if they're valid or not in your eyes. Help us understand.

**Comment by miguel-leon** [comment]
Sure, let me see... The simplest form of this feature would be:

- ✅ reassign1: `this` is `Foo`, `other` is `Foo`, `Foo` is assignable to `Foo`.
- ✅ reassign2: `other` is `Foo`, `this` is `Foo`, `Foo` is assignable to `Foo`.
- 🚫 reassign3: `this` is `Foo`, `other` is `SubFoo`, `Foo` is not assignable to `SubFoo`.
- ✅ reassign4: `other` is `SubFoo`, `this` is `Foo`, `SubFoo` is assignable to `Foo`.
- 🚫 reassign5: `this` is `Foo`, `other` is `IFoo`, `Foo` is not assignable to `IFoo`.
- ✅ reassign4: `other` is `IFoo`, `this` is `Foo`, `IFoo` is assignable to `Foo`.

In all theese cases, I'm assuming `Foo` has more stuff than `IFoo`, and `SubFoo` more stuff than `Foo`, which I'd imagine is your intent. In this particular example though, they're all structurally assignable to each other.

Was this what you were concerned with all along? I give you is a little tricky to reason about without confusing yourself, but it is actually intuitive.

The general rule is that a method in a superclass can always work as a method of the subclass, the opposite is not always true because the method in the subclass could expect more from `this`. Thus, you are allowed to assign a method of the supertype to the subtype. It's practically what happens upon inheritance.

**Comment by kirkwaiblinger** [comment]
@miguel-leon

> I am too in the real world you know 😅

😁 

> Sure, we can start a philosophical debate on software. I am aware, I assure you, that the most popular preference among programmers is to write things very "straightforward" or "beginner-friendly", specially in open source software or companies with large teams. And `if` statements that express intent so explicitly or literally, can certainly achieve that, even better would be `if (this.firstCallFlag) ... else ...`.
> 
> However, it is my believe that this preference is the result of a widespread intellectual laziness: If we're so (badly) accustomed of always relying on the same approach, it's no surprise we resist considering alternatives, and when presented, they inevitably seem alien or outrageous. Because of this, most programmers never develop the skills to quickly understand more complex code, relying on technologies in a thoughtless, robotic manner.
> 
> Returning to the example, I firmly believe that having less code to maintain—fewer entities like variables, methods and branches—pays off in the long run. And that's why I opt not to use flags, executors, setters, conditionals, run-time errors or any other approach that introduces unnecessary entities. I rather rely on the language's paradigm or capabilities, and javascript allows you to achieve the desired behavioral pattern with just the two methods `load` and `firstLoad`.

I hear you on this, but, with JS, it's not purely philosophical; there are _runtime_ differences in behavior. And IMO - the runtime behavior here is not desirable. Example later in response.

> > is a recipe for disaster to assign methods to non-methods
> > ```ts
> >  public notAMethod = this.method
> >  ...
> >  const unsafe = foo.notAMethod;
> > ```
> 
> Not really my case, and I don't understand why you would assign a method to a non-method without binding, but if you end up doing this, you can make it safe by typing it like below, and typescript would complain at the unsafe invokation.
> 
> ```ts
> public notAMethod: (this: Foo) => void = this.method
> ```
> 

Granted that the "Pass" example in the OP is given as N/A, we're left to take our best guess at what the request even is. To avoid confusion like this, you can help us out by giving us information to work with 🙏 . Which I see Brad has also requested in later comments.

> To conclude, just let me reiterate that I'm not doing any shenanigans here, the assignment is method to method, no intermediary untyped variables, the container object is the same type (it's even the same object), despite the `delete` the method is never even `undefined`, etc. This all look very sensible to me, so I'm guessing the analysis on all those degenerate edge cases is just for stress testing the suggestion, which is ok, but just only those examples and not mine, are the "bad code" ones.
> 
> I hope this clarifies any and all doubts. ✌️ Thanks again!

***

When assessing a request for a pattern that we're not familiar with, the questions that run through my head go something like this:
- does the pattern seem like it could be common?
   - my guess here is no. I've seen conditional code based on a boolean many times, never seen this.
- is the pattern easy to detect?
    - seemingly not. It's not been easy to ascertain what it even is from this conversation, which speaks to the degree of special-case logic involved 
- does the pattern seem like a good way of achieving its goal?
   - my answer here is no. It breaks inheritance, see below.
- are there alternatives?
   - Yes. Just store a boolean or write a sensible abstraction over one.

***

I think that shadowing a prototype method with an instance field which happens to be a method is undesirable because it breaks inheritance.

```ts
class Foo {
  constructor() {
    this.load = this.firstLoad;
  }

  private firstLoad() {
    // @ts-expect-error -- we're only deleting the instance own property, not the prototype method
    delete this.load;
    console.log('first load');
  }

  public load() {
    //...
  }
}

class Bar extends Foo {
  // This should be an error because you can't override an instance property with a prototype method.
  public override load() {
    console.log('overridden!')
  }
}

const bar = new Bar();
bar.load(); // expected: 'overridden', actual: 'first load'
```

Consider instead:

```ts
class Foo {
  private hasLoadBeenCalled = false;

  public load() {
    if (this.hasLoadBeenCalled) {
      throw new Error('already loaded!')
    }
    this.hasLoadBeenCalled = true;
    this.loadImpl();
  }

  private loadImpl() {
    console.log('base class load impl')
  }
}

class Bar extends Foo {
  override load() {
    console.log('derived class load impl');
  }
}

const bar = new Bar();
bar.load(); // 'derived class load impl', as expected
```

***

If the storing of a boolean bothers you.... abstract it!

```ts
function runOnce(f: () => void, afterFirstTime?: () => {}) {
  let isFirstRun = true;
  return () => {
    if (isFirstRun) {
      isFirstRun = false;
      f();
    } else {
      afterFirstTime?.()
    }
  }
}

class Foo {
  public load = runOnce(() => { this.loadImpl() });

  private loadImpl() {
    console.log('base class load impl')
  }
}

class Bar extends Foo {
  // Now this cannot be overridden by a prototype method, since it's an instance property of the parent class.
  // Fortunately, we get a TS error!
  //
  // TS error - 2425: Class 'Foo' defines instance member property 'load', but extended class 'Bar' defines it as instance member function.
  //
  // this error was previously disguised from TS by declaring the `this.load` method with a method, but immediately shadowing it.
  override load() {
    console.log('derived class load impl');
  }
}
```

***

In conclusion - since this pattern
1. does not seem idiomatic
2. has an idiomatic alternative
3. has unsoundness issues
4. has been difficult to describe straightforwardly enough to know how to implement it

I am -1.

If anything the more I think about it, it's surprising to me that TS even allows `this.prototypeMethod = somethingElse` at all. I think it should error.

**Comment by miguel-leon** [comment]
Sure @kirkwaiblinger, I see and understand your take on it. There's always going to be workarounds of the workarounds, and rebuttals of rebuttals, for example, one might say the intention should be to override the subsequent calls instead of the first one, in that case your example wouldn't be a problem at all. In other cases, there are also plenty of typing tweaks you can do to force errors on unsound usages.

@JoshuaKGoldberg, thanks for your opinion too. I don't consider solutions that involve so much more code and boilerplate cleaner solutions. But that's just how I program, and anyone should be free to program as they see fit. For that reason I don't find productive discussing on an issue things like "what would you do with it", "who's going to use it", "why this, not that". But I understand your point in that these questions are more for finding out whether to undertake the development or not.

I decided to report for two reasons:
Firstly because I saw that it might be simple enough to not take too much bandwidth (or I could even help if approved), and even supporting only most basic cases like the exact same type on both sides like @bradzacher would've been enough.
And secondly, because, patterns and usages aside, it is not semantically correct to say anything like "unbound-method rule error because the method is not bound", when it definitely is.

I feel like the report spiraled out of control somewhere, and the simplicity of the report was missed.
Thanks you guys anyway ✌️.
